crates/workers/src/router.rs

```diff
@@ -0,0 +1,253 @@
+                        self.router_name
+                    )));
+                };
+
+                model.chat_request(req).await
+            }
+            Some(worker::RouterConfig::Weighted { .. }) => {
+                let name = select_from_weighted(&self.models_cfg);
+                let Some(model) = self.models.read().await.get(&name).map(Arc::clone) else {
+                    return Err(OpenAIError::InvalidArgument(format!(
+                        "Model router '{}' expects a model '{name}' to exist, but does not",
+                        self.router_name
+                    )));
+                };
+
+                model.chat_request(req).await
+            }
+            Some(worker::RouterConfig::Fallback { .. }) => {
+                let fallbacks = into_ordered_fallbacks(&self.models_cfg);
+                for (name, _) in fallbacks {
+                    let Some(model) = self.models.read().await.get(&name).map(Arc::clone) else {
+                        return Err(OpenAIError::InvalidArgument(format!(
+                            "Model router '{}' expects a model '{name}' to exist, but does not",
+                            self.router_name
+                        )));
+                    };
+
+                    if let Ok(resp) = model.chat_request(req.clone()).await {
+                        return Ok(resp);
+                    }
+                }
+                Err(OpenAIError::ApiError(ApiError {
+                    message: format!(
+                        "All models in model router '{}' failed. Check logging for error details",
+                        self.router_name
+                    ),
+                    r#type: None,
+                    param: None,
+                    code: None,
+                }))
+            }
+            None => Err(OpenAIError::ApiError(ApiError {
+                message: format!("No models within model router '{}'.", self.router_name),
+                r#type: None,
+                param: None,
+                code: None,
+            })),
+        }
+    }
+
+    fn as_sql(&self) -> Option<&dyn SqlGeneration> {
+        None
+    }
+}
+
+/// Assumes all elements of `cfg` are [`worker::RouterConfig::Weighted`].
+/// Panics if no elements of `cfg` are [`worker::RouterConfig::Weighted`].
+fn select_from_weighted(cfg: &[worker::RouterConfig]) -> String {
+    let weighted: Vec<(String, u32)> = cfg
+        .iter()
+        .filter_map(|c| {
+            if let worker::RouterConfig::Weighted { from, weight } = c {
+                Some((from.clone(), *weight))
+            } else {
+                None
+            }
+        })
+        .collect::<Vec<_>>();
+
+    let index = if let Ok(dist) = WeightedIndex::new(weighted.iter().map(|(_, w)| w)) {
+        let mut rng = rng();
+        dist.sample(&mut rng)
+    } else {
+        0
+    };
+
+    weighted[index].0.clone()
```

Comment:
I don't think we are gauranteed that `weighted` has elements.

Reply:
Where/how it is currently used yes we are (comments atop the function), but may as well make it return Option<String> instead.